components/ui-checkbox/src/UiCheckbox.hooks.ts

```diff
@@ -0,0 +1,33 @@
+import { ChangeEventHandler, RefObject, useCallback, useEffect, useRef } from 'react';
+
+type Options = {
+  isIndeterminate: boolean;
+  onChange?: (isChecked: boolean) => void;
+};
+
+type Result = [RefObject<HTMLInputElement>, ChangeEventHandler<HTMLInputElement>];
+
+export function useLifecycle({ isIndeterminate, onChange }: Options): Result {
+  const ref = useRef<HTMLInputElement>(null);
+
+  // NOTE: Sets the `indeterminate` flag directly to the HTML element, because React doesn't support `indeterminate`
+  //       property natively.
+  useEffect(() => {
+    const { current: input } = ref;
+
+    if (input == null) {
+      return;
+    }
+
+    input.indeterminate = isIndeterminate;
+  }, [isIndeterminate]);
+
+  const handleChange: ChangeEventHandler<HTMLInputElement> = useCallback(
+    ({ target }) => {
+      onChange?.(target.checked);
+    },
+    [onChange],
+  );
+
+  return [ref, handleChange];
+}
```

Comment:
@amikhaylov-retable could you please, look at the hook? It's different compare to the main app implementation.

But actually, I couldn't understand for which purposes is used internal state in the app's implementation. Looks like a magic for some edgecases.

Reply:
As I remember, I added internal state to allow usage of single Checkbox without state. But, as I see in app, for now we don't use single Checkbox at all. Seems to me we can simplify and avoid this feature :)